### PR TITLE
raftstore: skip deleting when possible

### PR DESCRIPTION
**Pull request by BusyJay**
ref: https://github.com/pingcap/tikv/issues/1030

### CODE REVIEW

File: src/raftstore/store/engine.rs [hunk 1]
```diff
@@ -284,6 +284,10 @@ const MAX_DELETE_KEYS_COUNT: usize = 10000;
 /// finishes successfully but commit following `WriteBatch` failed, some keys are really deleted
 /// and can't be recovered.
 pub fn delete_all_in_range(db: &DB, start_key: &[u8], end_key: &[u8]) -> Result<()> {
+    if start_key >= end_key {
```
Comment (ngaut) [inline src/raftstore/store/engine.rs]: Is this a logical error?

Comment (BusyJay) [inline src/raftstore/store/engine.rs]: No, see peer_storage.rs#L513, `new_end_key` can be equal, less, or larger than `old_end_key`.

Comment (ngaut) [inline src/raftstore/store/engine.rs]: We need tests to cover the condition case.

Comment (BusyJay) [inline src/raftstore/store/engine.rs]: Existing cases in `test_snap.rs` and `test_split_region.rs` cover it already.

File: src/raftstore/store/peer_storage.rs [hunk 2]
```diff
@@ -503,6 +503,18 @@ impl PeerStorage {
         Ok(())
     }
 
+    /// Delete all data that is not covered by `new_region`.
+    fn clear_extra_data(&self, new_region: &metapb::Region) -> Result<()> {
+        let timer = Instant::now();
+        let (old_start_key, old_end_key) = (enc_start_key(self.get_region()),
+                                            enc_end_key(self.get_region()));
+        let (new_start_key, new_end_key) = (enc_start_key(new_region), enc_end_key(new_region));
+        try!(delete_all_in_range(&self.engine, &old_start_key, &new_start_key));
```
Comment (ngaut) [inline src/raftstore/store/peer_storage.rs]: What would happened if met the cross interval:
new_start_key.......old_start_key......new_end_key.....old_end_key
or 
new_start_key.......old_start_key.....old_end_key......new_end_key

Comment (BusyJay) [inline src/raftstore/store/peer_storage.rs]: It doesn't matter. Range `[new_start_key, new_end_key)` will be cleaned up when applying the snapshot data. So here just need to cleanup `[min(new_start_key, old_start_key), new_start_key)` and `[new_end_key, max(new_end_key, old_end_key))`.

Comment (ngaut) [inline src/raftstore/store/peer_storage.rs]: Gotcha. Thanks for explanation.

**Comment by ngaut** [comment]
LGTM

**Comment by siddontang** [comment]
LGTM

**Pull request merged**